File: view/android/VerificationCodeJavaDemo/app/src/main/java/com/example/verificationcodejavademo/widget/WordCaptchaDialog.java
package com.example.verificationcodejavademo.widget;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.Display;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.example.verificationcodejavademo.R;
import com.example.verificationcodejavademo.model.CaptchaCheckIt;
import com.example.verificationcodejavademo.model.WordCaptchaGetIt;
import com.example.verificationcodejavademo.network.BaseObserver;
import com.example.verificationcodejavademo.network.RetrofitUtils;
import com.example.verificationcodejavademo.network.RxHelper;
import com.example.verificationcodejavademo.utils.AESUtil;
import com.example.verificationcodejavademo.utils.ImageUtil;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import okhttp3.MediaType;
import okhttp3.RequestBody;

/**
 * Date:2020/5/19
 * author:wuyan
 */
public class WordCaptchaDialog extends Dialog {
    private String baseImageBase64;//背景图片
    private String token;
    private Context mContext;
    private TextView tvDelete;
    private ImageView tvRefresh;
    private WordImageView wordView;
    private TextView bottomTitle;
    private Handler handler = new Handler();
    private String key;


    public WordCaptchaDialog(@NonNull Context context) {
        super(context, R.style.dialog);
        this.mContext = context;
        setContentView(R.layout.dialog_word_captcha);
        getWindow().setGravity(Gravity.CENTER);
        getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        WindowManager windowManager = ((Activity) mContext).getWindowManager();
        Display display = windowManager.getDefaultDisplay();
        ViewGroup.LayoutParams lp = getWindow().getAttributes();
        lp.width = display.getWidth() * 9 / 10;
        getWindow().setAttributes((WindowManager.LayoutParams) lp);
        setCanceledOnTouchOutside(false);//点击外部Dialog不消失
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initView();
        loadCaptcha();

        tvDelete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });
        tvRefresh.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loadCaptcha();
            }
        });
    }

    private void initView() {
        tvDelete = findViewById(R.id.tv_delete);
        tvRefresh = findViewById(R.id.tv_refresh);
        wordView = findViewById(R.id.wordView);
        bottomTitle = findViewById(R.id.bottomTitle);

        Bitmap bitmap = ImageUtil.getBitmap(getContext(), R.drawable.bg_default);
        wordView.setUp(bitmap);
    }

    private void loadCaptcha() {
        bottomTitle.setText("数据加载中......");
        bottomTitle.setTextColor(Color.BLACK);
        Map<String, Object> params = new HashMap<>();
        params.put("captchaType", "clickWord");
        JSONObject jsonObject = new JSONObject(params);
        RequestBody body = RequestBody.create(MediaType.parse("application/json"), jsonObject.toString());
        Log.i("wuyan", "body:" + jsonObject);
        RetrofitUtils.getServerApi().getWordCaptchaAsync(body).compose(RxHelper.observableIO2Main(mContext)).subscribe(new BaseObserver<WordCaptchaGetIt>(mContext, true) {
            @Override
            public void onSuccess(WordCaptchaGetIt data) {
                baseImageBase64 = data.getOriginalImageBase64();
                token = data.getToken();
                key = data.getSecretKey();
                String wordStr = "";
                int j = 0;
                for (int i = 0; i < data.getWordList().size(); i++) {
                    j++;
                    wordStr += data.getWordList().get(i);
                    if (j < data.getWordList().size()) {
                        wordStr += ",";
                    }
                }
                wordView.setSize(data.getWordList().size());
                bottomTitle.setText("请依此点击【" + wordStr + "】");
                bottomTitle.setTextColor(Color.BLACK);
                wordView.setUp(
                        ImageUtil.base64ToBitmap(baseImageBase64)
                );
                initEvent();
            }

            @Override
            public void onFailure(Throwable e, String errorMsg) {
                bottomTitle.setText("加载失败,请刷新");
                bottomTitle.setTextColor(Color.RED);
                wordView.setSize(-1);
                Toast.makeText(mContext, errorMsg, Toast.LENGTH_SHORT).show();
            }
        });

    }

    private void checkCaptcha(String cryptedStr) {
        Map<String, Object> params = new HashMap<>();
        params.put("captchaType", "clickWord");
        params.put("token", token);
        params.put("pointJson", AESUtil.encode(cryptedStr, key));
        JSONObject jsonObject = new JSONObject(params);
        RequestBody body = RequestBody.create(MediaType.parse("application/json"), jsonObject.toString());
        Log.i("wuyan", "body:" + jsonObject);
        RetrofitUtils.getServerApi().checkAsync(body).compose(RxHelper.observableIO2Main(mContext)).subscribe(new BaseObserver<CaptchaCheckIt>(mContext, false) {
            @Override
            public void onSuccess(CaptchaCheckIt data) {
                bottomTitle.setText("验证成功");
                bottomTitle.setTextColor(Color.GREEN);
                wordView.ok();
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        loadCaptcha();
                        dismiss();
                    }
                }, 1000);
                if (mOnResultsListener!=null){
                    String result=token+"---"+cryptedStr;
                    mOnResultsListener.onResultsClick(AESUtil.encode(result,key));
                }
            }

            @Override
            public void onFailure(Throwable e, String errorMsg) {
                bottomTitle.setText("验证失败");
                bottomTitle.setTextColor(Color.RED);
                wordView.fail();
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        //刷新验证码
                        loadCaptcha();
                    }
                }, 1000);
            }
        });

    }

    private void initEvent() {
        wordView.setWordListenner(new WordImageView.WordListenner() {
            @Override
            public void onWord(String cryptedStr) {
                checkCaptcha(cryptedStr);
            }
        });
    }

    private OnResultsListener mOnResultsListener;

    public interface OnResultsListener {
        void onResultsClick(String result);
    }

    public void setOnResultsListener(OnResultsListener mOnResultsListener) {
        this.mOnResultsListener = mOnResultsListener;
    }
}

File: view/android/VerificationCodeDemo/app/src/main/java/com/example/verificationcodedemo/widget/WordCaptchaDialog.kt
package com.example.verificationcodedemo.widget

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.ViewGroup
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import com.example.verificationcodedemo.R
import com.example.verificationcodedemo.model.CaptchaCheckOt
import com.example.verificationcodedemo.model.CaptchaGetOt
import com.example.verificationcodedemo.network.Configuration
import com.example.verificationcodedemo.network.Configuration.token
import com.example.verificationcodedemo.utils.AESUtil
import com.example.verificationcodedemo.utils.ImageUtil
import kotlinx.android.synthetic.main.dialog_word_captcha.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

/**
 * Date:2020/5/8
 * author:wuyan
 */
class WordCaptchaDialog : Dialog {
    constructor(mContext: Context) : this(mContext, 0)
    constructor(mContext: Context, themeResId: Int) : super(
        mContext,
        com.example.verificationcodedemo.R.style.dialog
    ) {
        window!!.setGravity(Gravity.CENTER)
        window!!.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        val windowManager = (mContext as Activity).windowManager
        val display = windowManager.defaultDisplay
        val lp = window!!.attributes
        lp.width = display.width * 9 / 10//设置宽度为屏幕的0.9
        window!!.attributes = lp
        setCanceledOnTouchOutside(false)//点击外部Dialog不消失

    }

    var baseImageBase64: String = ""//背景图片
    var handler: Handler? = null
    var key: String = ""//ase加密秘钥


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(com.example.verificationcodedemo.R.layout.dialog_word_captcha)

        tv_delete.setOnClickListener {
            dismiss()
        }

        tv_refresh.setOnClickListener {
            wordView.reset()
            loadCaptcha()
        }

        //设置默认图片
        val bitmap: Bitmap = ImageUtil.getBitmap(context, R.drawable.bg_default)
        wordView.setUp(
            ImageUtil.base64ToBitmap(ImageUtil.bitmapToBase64(bitmap))!!
        )
        loadCaptcha()
    }

    private fun loadCaptcha() {
        Configuration.token = ""
        GlobalScope.launch(Dispatchers.Main) {
            try {
                bottomTitle.text = "数据加载中......"
                bottomTitle.setTextColor(Color.BLACK)
                wordView.visibility = View.INVISIBLE
                rl_pb_word.visibility = View.VISIBLE

                val o = CaptchaGetOt(
                    captchaType = "clickWord"
                )
                val b = Configuration.server.getWordCaptchaAsync(o).await().body()
                when (b?.repCode) {

                    "0000" -> {
                        baseImageBase64 = b.repData!!.originalImageBase64
                        Configuration.token = b.repData!!.token
                        key= b.repData!!.secretKey
                        var wordStr: String = ""
                        var i = 0;
                        b.repData!!.wordList!!.forEach {
                            i++
                            wordStr += it
                            if (i < b.repData!!.wordList!!.size)
                                wordStr += ","
                        }
                        wordView.setSize(b.repData!!.wordList!!.size)
                        bottomTitle.text = "请依此点击【" + wordStr + "】"
                        bottomTitle.setTextColor(Color.BLACK)
                        wordView.setUp(
                            ImageUtil.base64ToBitmap(baseImageBase64)!!
                        )
                        initEvent()
                    }
                    else -> {
                        bottomTitle.text = "加载失败,请刷新"
                        bottomTitle.setTextColor(Color.RED)
                        wordView.setSize(-1)
                    }
                }
                wordView.visibility = VISIBLE
                rl_pb_word.visibility = GONE

            } catch (e: Exception) {
                e.printStackTrace()
                Log.e("wuyan", e.toString())
                runUIDelayed(
                    Runnable {
                        bottomTitle.text = "加载失败,请刷新"
                        bottomTitle.setTextColor(Color.RED)
                        wordView.setSize(-1)
                        wordView.visibility = VISIBLE
                        rl_pb_word.visibility = GONE
                    }, 1000
                )
            }
        }
    }

    //检查验证码
    private fun checkCaptcha(pointListStr: String) {
        Log.e("wuyan", AESUtil.encode(pointListStr,key))
        GlobalScope.launch(Dispatchers.Main) {
            try {
                val o = CaptchaCheckOt(
                    captchaType = "clickWord",
                    pointJson = AESUtil.encode(pointListStr,key),
                    token = Configuration.token
                )
                val b = Configuration.server.checkAsync(o).await().body()
                when (b?.repCode) {

                    "0000" -> {
                        bottomTitle.text = "验证成功"
                        bottomTitle.setTextColor(Color.GREEN)
                        wordView.ok()
                        runUIDelayed(
                            Runnable {
                                dismiss()
                                loadCaptcha()
                            }, 2000
                        )

                        val result = token + "---" + pointListStr
                        mOnResultsListener!!.onResultsClick(AESUtil.encode(result, key))

                    }
                    else -> {
                        bottomTitle.text = "验证失败"
                        bottomTitle.setTextColor(Color.RED)
                        wordView.fail()
                        runUIDelayed(
                            Runnable {
                                //刷新验证码
                                loadCaptcha()
                            }, 1500
                        )
                    }
                }

            } catch (e: Exception) {
                e.printStackTrace()
                bottomTitle.text = "验证失败"
                bottomTitle.setTextColor(Color.RED)
                wordView.fail()
                runUIDelayed(
                    Runnable {
                        //刷新验证码
                        loadCaptcha()
                    }, 1500
                )
            }
        }
    }

    fun initEvent() {
        wordView.setWordListenner(object : WordImageView.WordListenner {
            override fun onWordClick(cryptedStr: String) {
                if (cryptedStr != null) {
                    checkCaptcha(cryptedStr!!)
                }
            }
        })
    }

    fun runUIDelayed(run: Runnable, de: Int) {
        if (handler == null)
            handler = Handler(Looper.getMainLooper())
        handler!!.postDelayed(run, de.toLong())
    }

    var mOnResultsListener: OnResultsListener? = null

    interface OnResultsListener {
        fun onResultsClick(result: String)
    }

    fun setOnResultsListener(mOnResultsListener: OnResultsListener) {
        this.mOnResultsListener = mOnResultsListener
    }


}